ledger-service/db-backend/src/test/scala/com/digitalasset/http/dbbackend/QueriesSpec.scala

```diff
@@ -59,6 +67,47 @@ class QueriesSpec extends AnyWordSpec with Matchers with TableDrivenPropertyChec
       )
     }
   }
+
+  "chunkBySetSize" should {
+    import org.scalacheck.{Arbitrary, Gen, Shrink}
+    import Arbitrary.arbitrary
+    import ScalaCheckDrivenPropertyChecks.{forAll => scForAll, _}
+    import Queries.chunkBySetSize
+
+    type Arg = NonEmpty[Map[Int, NonEmpty[Set[Int]]]]
+    val sizes: Gen[Int] = Gen.choose(1, 20)
+    implicit val doNotShrinkSize: Shrink[Int] = Shrink.shrinkAny
+    val randomArg: Gen[Arg] = Gen
+      .nonEmptyMap(
+        Gen.zip(
+          arbitrary[Int],
+          Gen.nonEmptyContainerOf[Set, Int](arbitrary[Int]).map { case NonEmpty(xs) => xs },
+        )
+      )
+      .map { case NonEmpty(xs) => xs }
+    implicit val shrinkArg: Shrink[Arg] = {
+      implicit val shrinkSet: Shrink[NonEmpty[Set[Int]]] =
+        Shrink { nes => Shrink.shrink(nes.forgetNE).collect { case NonEmpty(s) => s } }
+      Shrink { arg =>
+        Shrink.shrink(arg.forgetNE).collect { case NonEmpty(m) => m }
+      }
+    }
+
+    "include all arguments in the result" in scForAll(sizes, randomArg) { (s, r) =>
+      chunkBySetSize(s, r).foldMap1Opt(identity) should ===(Some(r))
+    }
+
+    def measuredChunkSize[K, V](chunk: NonEmpty[Map[K, NonEmpty[Set[V]]]]) =
+      chunk.toNEF.foldMap(_.size)
+
+    "never exceed size in each chunk" in scForAll(sizes, randomArg) { (s, r) =>
+      all(chunkBySetSize(s, r) map measuredChunkSize) should be <= s
+    }
+
+    "make chunks as large as possible" in scForAll(sizes, randomArg) { (s, r) =>
+      all(chunkBySetSize(s, r).init map measuredChunkSize) should ===(s)
+    }
+  }
```

Comment:
* [ ] extra property check: combining any two chunks with the map-of-intersection monoid should yield a map of all empty sets